### PR TITLE
Add widget for hiding resource previews.

### LINKED ISSUE #6542: A way to turn off the material preview in the Inspector?
**Operating system or device - Godot version:**
Godot 2.1, any OS.

**Issue description** (what happened, and what was expected):
_This is a feature request._

I don't really like the new shader material preview in the inspector, I am already previewing my materials in the viewport and it's kind of a waste of space in the inspector, IMO.

Illustration

Since I couldn't find an option for it:
Would it be possible to get an option to turn it off in the Engine Settings > Inspector Tab.

:sparkles: Pretty please? :sparkles:

### PR DESCRIPTION
**Pull request by GalanCM**
fixes #6542

Adds a new widget above resource previews to allow users to hide them.

Currently works for gradient, mesh, and texture resources. Would need to be implemented for sample and material previews as well, if they're ever re-implemented.

### CODE REVIEW

File: editor/resource_preview.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/*************************************************************************/
+/*  resource_preview.cpp                                            */
```
Comment (Noshyaar) [inline editor/resource_preview.cpp]: Please fix this line so that `*/` is aligned with others.

File: editor/resource_preview.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/*************************************************************************/
+/*  resource_preview.h                                              */
```
Comment (Noshyaar) [inline editor/resource_preview.h]: Please fix this one too.

**Comment by Noshyaar** [comment]
In my opinion:
1. The mouse cursor being up-down arrow when hovering suggests that it's draggable, but you interact with it by clicking.
2. The widget pushes the inspector up when switching dock tab. It could be a bit infuriating.
3. There being nothing like this widget anywhere else in the editor (toggling visibility with a splitter) and the widget being hidden until hovered might impact usability/discoverability.

File: editor/resource_preview.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	collapse_button->set_modulate(Color(0, 0, 0, 0));
+}
+
+void ResourcePreview::_notification(int p_what) {
+	if (p_what == NOTIFICATION_ENTER_TREE) {
+		collapse_button->set_normal_texture(get_icon("PanelCollapse", "EditorIcons"));
+		collapse_button->connect("toggled", this, "toggle_preview");
+		collapse_button->connect("mouse_entered", collapse_button, "set_modulate", varray(Color(1, 1, 1, 1)));
+		collapse_button->connect("mouse_exited", collapse_button, "set_modulate", varray(Color(0, 0, 0, 0)));
```
Comment (Noshyaar) [inline editor/resource_preview.cpp]: Are there any reasons behind `connect`ing these signals on enter tree? If there is, exit tree should disconnect them. If not, I think it should be moved to the contructor. The engine complains about connecting already connected signals when I move the inspector dock.

**Comment by GalanCM** [comment]
@Noshyaar I've fixed what I could from your suggestions. The reason for the visibility toggling and cursor change were because I was trying to mimic the behavior of the resize handles, since they're the most similar widget in Godot right now, but you're probably right about the discoverability problem. I changed that and it certainly feels better.

> The widget pushes the inspector up when switching dock tab. It could be a bit infuriating. (as the inspector's default location is lower dock)

I'm not quite sure what you mean here. Are you referring to the space that the widget takes up at the bottom when closed?

**Comment by Noshyaar** [comment]
Actually, it's not the widget that pushes it, it's the resource previewer. It's a bug in the current version. Sorry for misunderstanding. It'd be very nice if you can fix this too. :)

image

I like the recent changes. It's more obvious to me about what it does even though this is one of a few collapsible panels in the editor. Also, no more `connect` error spam. Thank you.

Finally, I'm not sure, but maybe the icons need dark variants as light theme is coming soon.

**Comment by GalanCM** [comment]
Alright, dark icons should be good. I couldn't replicate the pushing problem, even after merging from master, so I'm assuming it's been fixed.

**Comment by reduz** [comment]
After review it was decided that this will not be merged. After 3.0 the main focus is to remake the property editor to avoid these problems altogether, so making this area more complex to later remove it does no make much sense. Thanks for the effort though!

**Comment by kimjiy9607** [comment]
> After review it was decided that this will not be merged. After 3.0 the main focus is to remake the property editor to avoid these problems altogether, so making this area more complex to later remove it does no make much sense. Thanks for the effort though!

I was wondering if the separate property thing is under the progress currently.

**Pull request closed without merging**